Make Radicale fast

Unrud opened the pull request:
Big calendars/addressbooks are really slow and requests require a lot of memory. The main reason is vobject. Parsing and serializing with it is slow and it's a memory hog.
This PR reduces the usage of vobject by caching serialization results and metadata for pre-filtering on disk. The memory usage gets further reduced by only keeping one item in memory at a time.

All commits until ``Cache metadata when storage is read-only`` are general improvements. They fix some small bugs, reduce the memory usage by not loading all items at once into memory and fix the slowness that was introduced by collision detection for file system paths.
The following commits improve speed by:
  * Caching the serialization of the whole collection on disk.
  * Caching the serialization of items and metadata (tag, start and end time) for pre-filtering on disk.
  * Caching the etag and metadata of the collection for the duration of the request in memory, if the access is read-only.

This PR is also merged into https://github.com/Unrud/Radicale.

## Benchmark
The tests are performed on a calendar and addressbook with 10,000 files. The approximate memory usage of the requests is noted in brackets.

### Calendar
```
Type:   PROPFIND_getetag (initialize cache)
Before: 15.5 seconds (138,156 kB)
After:  17.8 seconds (    372 kB)

Type:   PROPFIND_getetag
Before: 15.3 seconds (136,604 kB)
After:   0.5 seconds (      0 kB)

Type:   REPORT_calendar-query_getetag
Before: 16.8 seconds (154,180 kB)
After:   1.0 seconds ( 21,964 kB)
PR #586: 1.0 seconds ( 15,356 kB)

Type:   REPORT_calendar-multiget_getetag
Before: 16.4 seconds ( 28,248 kB)
After:   2.1 seconds ( 28,660 kB)
PR #586: 2.2 seconds ( 19,852 kB)

Type:   REPORT_calendar-query_filter-all_getetag
Before: 18.3 seconds (154,448 kB)
After:   1.1 seconds ( 22,332 kB)
PR #586: 1.1 seconds ( 15,092 kB)

Type:   REPORT_calendar-query_filter-none_getetag
Before: 11.0 seconds (133,580 kB)
After:   0.5 seconds (      0 kB)

Type:   GET
Before: 30.5 seconds (138,756 kB)
After:   1.2 seconds (  2,740 kB)
```

### Addressbook
```
Type:   PROPFIND_getetag (initialize cache)
Before:  8.0 seconds ( 69,732 kB)
After:   9.4 seconds (    284 kB)

Type:   PROPFIND_getetag
Before:  8.0 seconds ( 70,020 kB)
After:   0.5 seconds (      0 kB)

Type:   GET
Before: 16.0 seconds ( 70,012 kB)
After:   1.1 seconds (  3,692 kB)
```
**Note:** In all "Before" tests the collision checking for file names was disabled, because it slowed down the requests significantly and made the results less comparable.

pbiering commented:
sounds interesting and probably superseed my "simple" in-memory caching, yours is persistent and survives Radicale restarts.

Unrud commented:
I've implemented prefiltering for the multifilesystem backend and all collection items are now verified by vobject. For this I added a on-disk cache that stores the etags of verified items and some extracted parameter for prefiltering. Items that are not in the cache are always parsed and serialized before they are returned by **Collection.get**.

Prefiltering takes about 1 second to filter out 10,000 items.

I've added some helper functions for prefiltering to xmlutils. They convert the complicated XML filters to a tag and time range (POSIX timestamps as ints) and can find the tag and an enclosing time range for vobject_items. 

> superseed my "simple" in-memory caching

I did some experiments. I loaded a collection (with simple VEVENTs) that required ~3.4 MiB of space on disk. Caching it in memory required an additional ~128 MiB of RAM.

Unrud commented:
I changed the item cache to only keep one cache entry per item. Entries for deleted items get cleaned up automatically.

Unrud commented:
Rebased, fixed all flake8 warnings, simplified and reordered the commits and added some comments.

All commits until ``Cache metadata when storage is read-only`` are general improvements. They fix some small, reduce the RAM usage by not loading all items at once into memory and fix the slowness that was introduced by collision detection for file system paths.
The following commits improve speed by:
  * Caching the serialization of the whole collection on disk.
  * Caching the serialization of items and metadata (tag, start and end time) for pre-filtering on disk.
  * Caching the etag and metadata of the collection for the duration of the request in memory, if the access is read-only.

Unrud commented:
I ran the benchmark again, to incorporate the new changes. This time I also included the approximate RAM usage of the requests. The tests are performed on a calendar and addressbook with 10,000 files.

## Benchmark
### Calendar
```
Type:   PROPFIND_getetag (initialize cache)
Before: 15.5 seconds (138,156 kB)
After:  17.8 seconds (    372 kB)

Type:   PROPFIND_getetag
Before: 15.3 seconds (136,604 kB)
After:   0.5 seconds (      0 kB)

Type:   REPORT_calendar-query_getetag
Before: 16.8 seconds (154,180 kB)
After:   1.0 seconds ( 21,964 kB)

Type:   REPORT_calendar-multiget_getetag
Before: 16.4 seconds ( 28,248 kB)
After:   2.1 seconds ( 28,660 kB)

Type:   REPORT_calendar-query_filter-all_getetag
Before: 18.3 seconds (154,448 kB)
After:  10.2 seconds ( 22,332 kB)

Type:   REPORT_calendar-query_filter-none_getetag
Before: 11.0 seconds (133,580 kB)
After:   0.5 seconds (      0 kB)

Type:   GET (initialize cache)
Before: 30.4 seconds (138,752 kB)
After:  16.3 seconds (102,228 kB)

Type:   GET
Before: 30.5 seconds (138,756 kB)
After:   0.6 seconds (  6,860 kB)
```

### Addressbook
```
Type:   PROPFIND_getetag (initialize cache)
Before:  8.0 seconds ( 69,732 kB)
After:   9.4 seconds (    284 kB)

Type:   PROPFIND_getetag
Before:  8.0 seconds ( 70,020 kB)
After:   0.5 seconds (      0 kB)

Type:   GET (initialize cache)
Before: 16.0 seconds ( 70,012 kB)
After:   1.1 seconds (  3,836 kB)

Type:   GET
Before: 15.8 seconds ( 69,460 kB)
After:   0.6 seconds (  5,012 kB)
```
**Note:** In all "Before" tests the collision checking for file names was disabled, because it slowed down the requests significantly and made the results less comparable.

Unrud commented:
I implemented the serialization of calendars without vobject. The calendar components are merged manually. This improves the time of GET requests from ~30 seconds (~100 MiB of memory) to ~1 second (~5 MiB of memory). Consequentially I removed the cache for the serialization of collections, because it's not needed anymore.

Filtering of items with vobject in ``xmlutils.py`` can now be skipped if the storage backend indicates that it already matched all ``filters`` in prefiltering. This is used to optimize requests with simple filters, that only match the tag and a big time range (like filtering all events from the last three months).
At least DAVdroid and InfCloud only make use of simple filters. This improves the time for the requests from ~10 seconds to ~1 second.

vobject is now only used for one-time input validation and for complex filtering.
I've updated the benchmark in the first post and also included results from #586.

I've created a fork of Radicale at https://github.com/Unrud/Radicale where I merged all my open PRs.

jvoisin commented:
Nice!
What is needed to get this merged @liZe ?

htgoebel commented on radicale/storage.py:
```diff
@@ -687,53 +862,87 @@ def delete(self, href=None):
+                                    vtimezones += "\r\n"
+                                vtimezones += "\r\n".join(vtimezone)
+                                included_tzids.add(tzid)
+                            vtimezone.clear()
```
Why not simply `vtimezone = []`, which would be more obfious?

Unrud replied:
Because it uses more memory (the GC doesn't cleanup the old memory immediately) and it's slower. This makes a big difference.

htgoebel commented on radicale/storage.py:
```diff
@@ -687,53 +862,87 @@ def delete(self, href=None):
+            vtimezone = []
+            tzid = None
+            components = ""
+            for item in self.get_all():
```
Please add a comment what this code does and why.

htgoebel commented on radicale/storage.py:
```diff
@@ -687,53 +862,87 @@ def delete(self, href=None):
+                    elif vtimezone:
+                        vtimezone.append(line)
+                        if depth == 2 and line.startswith("TZID:"):
+                            tzid = line[len("TZID:"):]
```
Use a constant here : `line = line[5:]`

Unrud replied:
I tried to avoid magic numbers.

htgoebel commented on radicale/storage.py:
```diff
@@ -687,53 +862,87 @@ def delete(self, href=None):
+                        vtimezone.append(line)
+                    elif vtimezone:
+                        vtimezone.append(line)
+                        if depth == 2 and line.startswith("TZID:"):
```
This would also match if `TZID:` is part of another nested structure. Is this indented?

Unrud replied:
It doesn't because of ``depth == 2``.

htgoebel replied:
I mean this case:
```
BEGIN:Rubbish
BEGIN:VTIMEZONE
TZID:
…
```

htgoebel replied:
I mean this case (which *may* be an invalid vevent, but this *may* change in some revision of the standard)
```
BEGIN:Rubbish
BEGIN:VTIMEZONE
TZID:
…
```

htgoebel commented on radicale/storage.py:
```diff
@@ -687,53 +862,87 @@ def delete(self, href=None):
+            etag = md5()
+            for item in self.get_all():
+                etag.update((item.href + "/" + item.etag).encode("utf-8"))
+            self._etag = '"%s"' % etag.hexdigest()
```
Using string-concatenation (`+`) is more obvious and should be as fast.

Unrud replied:
> Using string-concatenation

I used this method, because it's also used at other places in the code. I don't think that it matters.

> should be as fast.

This is only done once, speed is not a concern.

htgoebel commented on radicale/xmlutils.py:
```diff
@@ -207,11 +209,9 @@ def _prop_match(item, filter_):
 
 def _time_range_match(vobject_item, filter_, child_name):
-    """Check whether the ``item`` matches the time-range ``filter_``.
+    """Check whether the the component/property ``child_name`` of
```
typo

Unrud replied:
Yes.

htgoebel commented on radicale/xmlutils.py:
```diff
@@ -226,14 +226,58 @@ def _time_range_match(vobject_item, filter_, child_name):
+    DAY = timedelta(days=1)
+    SECOND = timedelta(seconds=1)
+    DATETIME_MIN = datetime.min.replace(tzinfo=timezone.utc)
+    DATETIME_MAX = datetime.max.replace(tzinfo=timezone.utc)
```
These are constant values and should be global.

Unrud replied:
Yes.

htgoebel commented on radicale/xmlutils.py:
```diff
@@ -426,6 +493,107 @@ def _param_filter_match(vobject_item, filter_, parent_name):
+    TIMESTAMP_MIN = math.floor(
+        datetime.min.replace(tzinfo=timezone.utc).timestamp())
+    TIMESTAMP_MAX = math.ceil(
+        datetime.max.replace(tzinfo=timezone.utc).timestamp())
```
These are constants and should be global

Unrud replied:
Yes.

htgoebel commented on radicale/xmlutils.py:
```diff
@@ -426,6 +493,107 @@ def _param_filter_match(vobject_item, filter_, parent_name):
+    """
+
+    DATETIME_MIN = datetime.min.replace(tzinfo=timezone.utc)
+    DATETIME_MAX = datetime.max.replace(tzinfo=timezone.utc)
```
dito.

Unrud replied:
Yes.

Unrud commented on radicale/storage.py:
```diff
@@ -207,19 +237,42 @@ def __init__(self, path):
 class Item:
-    def __init__(self, collection, item, href, last_modified=None):
+    def __init__(self, collection, href, last_modified=None,
+                 text=None, item=None, etag=None):
```
> Please add a doc-string

Yes.

> The interface changed.

Totally missed that. I also changed the behavior of ``pre_filtered_list``.

> This should go into the release notes.

Or I reorder the parameters and make it backwards compatible. It depends, if this PR gets merged before the 2.0.0 release.

Unrud commented on radicale/storage.py:
```diff
@@ -207,19 +237,42 @@ def __init__(self, path):
-    def __init__(self, collection, item, href, last_modified=None):
+    def __init__(self, collection, href, last_modified=None,
+                 text=None, item=None, etag=None):
+        if text is None and item is None:
```
``text`` is just the raw data from the storage. It will fail as soon as it's parsed with **vobject**, like any other invalid data. I don't think that we should introduce a special case for one kind of invalid data. That would also mean, that we have to duplicate the logging.

Unrud commented on radicale/storage.py:
```diff
@@ -207,19 +237,42 @@ def __init__(self, path):
+    def __init__(self, collection, href, last_modified=None,
+                 text=None, item=None, etag=None):
+        if text is None and item is None:
+            raise ValueError("text and item are not set")
```
Yes.

Unrud commented on radicale/storage.py:
```diff
@@ -304,23 +357,40 @@ def get(self, href):
+        exist.
+
+        """
+        return map(lambda href: (href, self.get(href)), hrefs)
```
No, that would defeat the main purpose of this PR. A generator expression will load everything into memory at once. Avoiding generator expressions is the main reason, why this PR reduces the memory usage significantly.

htgoebel replied:
I'm curios about this. AFAIK (and I'm quite sure about this) generator expressions (using parents `(…)`) are like iterators. Opposed to list comprehension (using brackets `[…]`) which are like lists. And iterators are not loading all data into memory at once. But I may be wrong and am really curious to learn why.

Unrud replied:
You are right, I I mixed them up.

Unrud commented on radicale/storage.py:
```diff
@@ -304,23 +357,40 @@ def get(self, href):
         """
-        for href in set(hrefs):
-            yield self.get(href)
+        return map(self.get, self.list())
```
No, see comment about generator expressions.

Unrud commented on radicale/storage.py:
```diff
@@ -304,23 +357,40 @@ def get(self, href):
         This returns all event by default
         """
-        return [self.get(href) for href in self.list()]
+        return map(lambda item: (item, False), self.get_all())
```
No, see comment about generator expressions.

Unrud commented on radicale/storage.py:
```diff
@@ -621,43 +689,150 @@ def move(cls, item, to_collection, to_href):
+                os.remove(os.path.join(folder, name))
+            except (FileNotFoundError, PermissionError):
+                continue
+            modified = True
```
> Shorter and more focused on on what should be done here

It's not equivalent. The file is not removed if ``age_limit`` is ``None``.

> compute path only once for speed

Maybe, but it shouldn't make a measurable difference.

Unrud commented on radicale/storage.py:
```diff
@@ -621,43 +689,150 @@ def move(cls, item, to_collection, to_href):
+            except PermissionError:
+                pass
+            # Clean cache entries (max once per request)
+            if not self._item_cache_cleaned:
```
This only happens once, after new items are uploaded to the storage. I put it here and not into ``def upload(...)`` because the user can also edit the file system storage by hand.

Unrud commented on radicale/storage.py:
```diff
@@ -621,43 +689,150 @@ def move(cls, item, to_collection, to_href):
+                        os.path.join(self._filesystem_path, href))
+
+                self._clean_cache(cache_folder, filter(
+                    is_invalid_href, scandir(cache_folder)))
```
No, see comment about generator expressions. Although in this case it's a bit overkill, this doesn't save any significant amount of memory. Directory contents are usually not that big.

Unrud commented on radicale/storage.py:
```diff
@@ -621,43 +689,150 @@ def move(cls, item, to_collection, to_href):
-            raise
-        return Item(self, item, href, last_modified)
+        return Item(self, href, last_modified=last_modified, etag=cetag,
+                    text=ctext, item=vobject_item), (ctag, cstart, cend)
```
You can't return ``vobject_item`` directly. The ``text`` is the raw text from the file system. You have to serialize it again, to normalize the text with **vobject**. The user might have edited the storage in the file system by hand, it might not be normalized.

Unrud commented on radicale/storage.py:
```diff
@@ -621,43 +689,150 @@ def move(cls, item, to_collection, to_href):
+        files = None
+        for href in hrefs:
+            if files is None:
+                # Only list dir when hrefs is not empty
```
That is alright. If the ``hrefs`` iterator doesn't contain anything, the for-loop is never executed.
Therefore, we only list the directory content if the ``hrefs`` iterator is not empty. That's the whole purpose of this.

Unfortunately it's somewhat complicated to check if the iterator is empty, before the for-loop.

htgoebel replied:
Puh, this is tricky :-) I suggest adding a comment so other nitpickers keep quiet.

Unrud commented on radicale/storage.py:
```diff
@@ -621,43 +689,150 @@ def move(cls, item, to_collection, to_href):
+    def get_all(self):
+        # We don't need to check for collissions, because the the file names
+        # are from os.listdir.
+        return map(lambda x: self.get(x, verify_href=False), self.list())
```
No, see comment about generator expressions.

Unrud commented on radicale/storage.py:
```diff
@@ -621,43 +689,150 @@ def move(cls, item, to_collection, to_href):
+        if not tag:
+            # no filter
+            yield from map(lambda item: (item, simple), self.get_all())
+            return
```
I tried to prevent deep indentations.

Unrud commented on radicale/storage.py:
```diff
@@ -621,43 +689,150 @@ def move(cls, item, to_collection, to_href):
+            return
+        for item, (itag, istart, iend) in map(
+                lambda x: self._get_with_metadata(x, verify_href=False),
+                self.list()):
```
No, see comment about generator expressions.

Unrud commented on radicale/storage.py:
```diff
@@ -687,53 +862,87 @@ def delete(self, href=None):
+        new_props = self.get_meta()
+        new_props.update(props)
+        for key in tuple(new_props.keys()):
+            if not new_props[key]:
```
It's the same behavior as the old code. ``xmlutils`` uses it to delete entries.

Unrud commented:
@htgoebel Thanks for reviewing, I will fix the identified problems soon.
